multisig/src/util.rs

```diff
@@ -1,40 +1,107 @@
 use std::ops::Deref;
 
-use ed25519_compact as ed25519;
+use ed25519_compact::x25519;
 use serde::{Deserialize, Deserializer, Serialize, Serializer, de};
 use serde_bytes::ByteArray;
 
+use crate::InvalidSecretKey;
+
 pub(crate) fn encode<S, T, const N: usize>(d: &T, s: S) -> Result<S::Ok, S::Error>
 where
     S: Serializer,
     T: Deref<Target = [u8; N]>,
 {
-    ByteArray::new(**d).serialize(s)
+    if s.is_human_readable() {
+        bs58::encode(**d).into_string().serialize(s)
+    } else {
+        ByteArray::new(**d).serialize(s)
+    }
 }
 
-pub(crate) fn decode_pk<'de, D>(d: D) -> Result<ed25519::PublicKey, D::Error>
+pub(crate) fn decode_x25519_pk<'de, D>(d: D) -> Result<x25519::PublicKey, D::Error>
 where
     D: Deserializer<'de>,
 {
-    let a = ByteArray::deserialize(d)?.into_array();
-    if super::SMALL_ORDER_KEYS.contains(&&a) {
-        return Err(de::Error::custom("small order public key detected"));
+    if d.is_human_readable() {
+        let s = String::deserialize(d)?;
+        let mut a = [0; 32];
+        let n = bs58::decode(&s).onto(&mut a).map_err(de::Error::custom)?;
+        x25519::PublicKey::from_slice(&a[..n]).map_err(de::Error::custom)
+    } else {
+        let a = ByteArray::<32>::deserialize(d)?;
+        x25519::PublicKey::from_slice(&a[..]).map_err(de::Error::custom)
+    }
+}
+
+pub(crate) fn decode_x25519_sk<'de, D>(d: D) -> Result<x25519::SecretKey, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    if d.is_human_readable() {
+        let s = String::deserialize(d)?;
+        let mut a = [0; 32];
+        let n = bs58::decode(&s).onto(&mut a).map_err(de::Error::custom)?;
+        x25519::SecretKey::from_slice(&a[..n]).map_err(de::Error::custom)
+    } else {
+        let a = ByteArray::<32>::deserialize(d)?;
+        x25519::SecretKey::from_slice(&a[..]).map_err(de::Error::custom)
+    }
+}
```

Comment:
i'm unsure about our serde strategy, esp why do we have to convert these objects into ByteArray first? 

the "per-object" serde functions also kinda of baffles me.

Reply:
This is a good point. I think for the non-human readable format this is not required.

Reply:
Actually the `Serialize` and `Deserialize` implementations of the various `secp256k1` types use the regular sequential forms of serde and here, similar to the implementations for the x25519 keys, we use the optimised `serde_bytes` form. This results for example in a single `Serializer::serialize_bytes` call instead of serializing the key as a tuple and then each element.